common/flatpak-run.c

```diff
@@ -3165,6 +3169,88 @@ regenerate_ld_cache (GPtrArray    *base_argv_array,
   return glnx_steal_fd (&ld_so_fd);
 }
 
+/* Check that this user is actually allowed to run this app. When running
+ * from the gnome-initial-setup session, an app filter might not be available. */
+static gboolean
+check_parental_controls (const char     *app_ref,
+                         FlatpakDeploy  *deploy,
+                         GCancellable   *cancellable,
+                         GError        **error)
+{
+#ifdef HAVE_LIBMALCONTENT
+  g_auto(GStrv) app_ref_parts = NULL;
+  g_autoptr(MctManager) manager = NULL;
+  g_autoptr(MctAppFilter) app_filter = NULL;
+  g_autoptr(GAsyncResult) app_filter_result = NULL;
+  g_autoptr(GDBusConnection) system_bus = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  app_ref_parts = flatpak_decompose_ref (app_ref, error);
+  if (app_ref_parts == NULL)
+    return FALSE;
+
+  system_bus = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, error);
+  if (system_bus == NULL)
+    return FALSE;
+
+  manager = mct_manager_new (system_bus);
+  app_filter = mct_manager_get_app_filter (manager, getuid (),
+                                           MCT_GET_APP_FILTER_FLAGS_INTERACTIVE,
+                                           cancellable, &local_error);
+  if (local_error != NULL &&
+      !g_error_matches (local_error, MCT_APP_FILTER_ERROR, MCT_APP_FILTER_ERROR_INVALID_USER))
+    {
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return FALSE;
+    }
+  g_clear_error (&local_error);
+
+  if (app_filter != NULL)
+    {
+      g_autoptr(GDesktopAppInfo) app_info = NULL;
+      gboolean allowed = FALSE;
+
+      if (deploy != NULL)
+        {
+          g_autoptr(GFile) deploy_dir = NULL;
+          const char *deploy_path;
+          g_autofree char *desktop_file_name = NULL;
+          g_autofree char *desktop_file_path = NULL;
+
+          deploy_dir = flatpak_deploy_get_dir (deploy);
+          deploy_path = flatpak_file_get_path_cached (deploy_dir);
+
+          desktop_file_name = g_strconcat (app_ref_parts[1], ".desktop", NULL);
+          desktop_file_path = g_build_path (G_DIR_SEPARATOR_S,
+                                            deploy_path,
+                                            "export",
+                                            "share",
+                                            "applications",
+                                            desktop_file_name,
+                                            NULL);
```

Comment:
I don't think this is necessarily a huge problem, but just a note here. An app can have multiple desktop files, but this is only using the main one. This means all desktop files in an app will have the same permissions, and that all info is taken from the main one only.

Reply:
Ah, good point. I think that’s OK though? Primarily, we expect flatpaks to be blacklisted based on their app ID, which should be present in all the `.desktop` files in the flatpak, right?

To be a bit safer, I’ll tweak it so that it always does the app ref check, rather than only doing it if it can’t do the app info check.

`mct_app_filter_is_appinfo_allowed()` is basically a convenience wrapper around several checks, which happen to all be possible with the information provided by the `GAppInfo`: https://gitlab.freedesktop.org/pwithnall/malcontent/blob/master/libmalcontent/app-filter.c#L273

Looking at the code which actually chooses which binary to run, it looks like that’s decoupled from what `.desktop` files are actually in the flatpak. Perhaps we need to add a metadata key in `flatpak-builtins-build-finish.c` which stores which `.desktop` file is the ‘main’ one associated with the command stored in `FLATPAK_METADATA_KEY_COMMAND`?

Reply:
Actually there doesn't necessarily have to be a match of what "flatpak run" and the default desktop file runs. In practice they are probably the same, but they are really independent.

Reply:
So perhaps we should set some metadata saying what the ‘main’ desktop file is, if and only if that corresponds to the `FLATPAK_METADATA_KEY_COMMAND`. If no such metadata is set, `flatpak run` would not check any desktop file.